### ISSUE TITLE
Routines need to be able to manipulate strings of numerals

### ISSUE THREAD

**Issue by robartsd** [ArnoldSmith86/virtualtabletop#564]
Currently computations ALWAYS interpret strings of numerals as numbers. There needs to be ways to explicitly work with them as strings.

**Comment by robartsd** [comment]
An automatically generated `id` happened to be all numerals. 'id` placed in array by a routine and stored in `cardsPlayed` property. Stored property contained number instead of string. This select fails to capture the card.
```json
    {
      "func": "SELECT",
      "property": "id",
      "relation": "in",
      "value": "${PROPERTY cardsPlayed}"
    }
```

**Comment by robartsd** [comment]
I think the toNum conversion should be moved to be operation sensitive:

Should not convert to number:
- `+` (unless both operands can be converted)
- Alphanumeric Comparisons (unless both operands can be converted)
- arguments to string and array functions that are not explicitly numeric
- object indexes

**Comment by ArnoldSmith86** [comment]
Changing that will be really ugly. The fileupdater would have to add it before nearly every operation...

**Comment by robartsd** [comment]
> Changing that will be really ugly. The fileupdater would have to add it before nearly every operation...

It needs to be fixed even if it breaks things.

**Comment by ArnoldSmith86** [comment]
I don't agree. It should be fixed (or added) but we shouldn't break things.

**Comment by robartsd** [comment]
> I don't agree. It should be fixed (or added) but we shouldn't break things.

It needs to be fixed. If it can be fixed without breaking things that is great. If it can't be fixed without breaking things, then things break.

**Comment by robartsd** [comment]
A possible fix that would not break things would be to add some way to switch mode between converting strings to numbers or not (defaulting to current behavior or using fileUpdater to override the new default).

**Comment by ArnoldSmith86** [comment]
I mean, sure. But it can 100% be done without breaking stuff. The easiest solution would be to have a whole different set of operations.

A mode switch is already 100x better and maybe we find an even better solution.

**Comment by robartsd** [comment]
> I mean, sure. But it can 100% be done without breaking stuff. The easiest solution would be to have a whole different set of operations.
> 
> A mode switch is already 100x better and maybe we find an even better solution.

I'd be okay if a mode switch turns off conversion to numbers completely, but I'd prefer conversion to number continue regardless of mode switch anywhere only a number would make sense.

Fortunately conversion from number to string happens automatically in many cases.

**Comment by robartsd** [comment]
Mode switch in Dynamic Expressions will be a small pain to write, simply because many (if not all) of the capturing indexes will change. Too bad named capture groups are probably not well enough supported for us to simply go with named capture groups as a more maintainable solution.

**Comment by ArnoldSmith86** [comment]
Yeah. I think we should list all the functions that would need a toggle (or always disabled). Then we add a mode toggle that defaults to "no conversion for those functions" and implement a fileupdater that adds the toggle at the start of every routine that uses one of those functions.

**Comment by robartsd** [comment]
Would we toggle at the routine level or the routine step level? At first I was thinking a flag in every step, but a step that changes the options might be nicer.

```JSON
"...Routine": [
  "toNum: true",
  "var x = '100' // x gets set to 100",
  "toNum: false",
  "var y = '100' // y gets set to '100'"
}
```

**Comment by robartsd** [comment]
We discussed in our meeting today about adding a mode switch to routines. To accommodate future changes we want a format that allows setting multiple modes in one line.
```json
"...Routine": [
   "modes: autoConvertToNumber"
   ...current_behavior...
]
```
Another mode we mentioned would be for legacy setting default values of operands to `1`.
